README.md
Spinsim 0.75

This version of spinsim supports about two-thirds of the opcodes in the P2
instruction set.  The list of implemented opcodes is shown below, along with
the list of opcodes that have not been implemented yet.  Spinsim runs in the P1
mode by default.  It can be set to run in the P2 mode with the -t parameter.

In the P2 mode, spinsim will load a P2 OBJ file into hub RAM, and start cog 0
with the code located at $E00.  A simulated serial port is supported by using
the -b option.  The default baud rate is 115,200, but other rates can be used
by specifying it with -b, such as -b9600.  The serial port uses pin 31 and 30,
just like with P1.

Spinsim is built using the Makefile, and typing make.  I have successfully
built and run it under Cygwin and Ubuntu Linux.

The sample program, pfth.spin can be run by building it with the PNut P2
assembler.  PNut is contained in the Terasic_Prop2_Emulation zip file,
which can be downloaded from the first post in the "HUB EXEC Update Here"
thread in the Propeller 2 forum.

There are two demo programs, which are the pfth Forth interpreter and the
p1spin Spin interpreter that runs P1 Spin binaries on the P2 processor.
The pfth program is run under spinsim as follows:

./spinsim -t -b pfth.obj

p1spin runs at a baud rate of 57600, so it is run as follows:

./spinsim -t -b57600 p1spin.obj

Spinsim supports execution from cog memory and hub execution, but it does not
support multi-tasking.  Only the core processor is supported, and none of the
counters, video hardware or cordic hardware is simulated.  Support for multi-
tasking, peripheral hardware and cordic instructions will be added later.

Spinsim contains a simple debugger, which is enabled with the -d command-line
option.  The debugger prints the prompt "DEBUG>" to indicate that it is ready
to accept a command.  The "help" command will print the following:

Debug Commands
help           - Print command list
exit           - Exit spinsim
step           - Run one cycle
stepx          - Run next executed instruction
run            - Run continuously
verbose #      - Set verbosity level
reboot         - Reboot the Prop
setbr cog addr - Set breakpoint for cog to addr
state cog      - Dump cog state

The "step" command will run one cycle, and the "stepx" command will run any
non-executing cycles until it encounters an instruction that is executed.
The previous command can be executed again by just pushing the enter key.
This is useful for stepping where the "step" command is typed once, and
the enter key can then be used to step again.

The "run" command will run until a breakpoint is encountered or ^] is typed.
^] is typed by holding down the control key and pressing the "]" key.
While running, spinsim will print out the results of each cycle as
controlled by the verbosity level, which is set by the "verbose" command.
The verbosity level can also be set with the command-line parameter "-v#".

The verbosity levels are as follows:

0 - Disable printing
1 - Print only the executed instructions
2 - Print only the executed instructions, and show the execution results
3 - Print executed and instruction not executed due to condition code
4 - Also print instructions invalidated in the pipeline due to jumps
5 - Also print instructions that include hub and hardware waits
6 - Also print instructions that include icache waits
7 - Also print instructions waiting for a pin state
8 - Print all cycles, including waitcnt waits

The verbosity level is entered as a hexadecimal number.  If the verbosity level
is entered as a single digit it will apply to all cogs.  If more than one digit
is entered each digit will be used for each cog, starting with cog 0 for the
right-most digit.  As an example, a value of 456 will cause 6 to be used for
cog 0, 5 for cog 1, and 4 for cog 2.  All other cogs will use 0.


Implemented Opcodes
-------------------
abs     add     addabs  addptra addptrb addptrx addptry adds
addsx   addx    and     andn    augd    augs    bingry  blmask
call    calla   callad  callb   callbd  calld   callx   callxd
cally   callyd  chkptrx chkptry clkset  clracca clraccb clraccs
clrb    clrp    cmp     cmpcnt  cmps    cmpsub  cmpsx   cmpx
cogid   coginit cognew  cogstop dcachex decd    decds   decmod
decod2  decod3  decod4  decod5  div32   div32u  div64   div64d
div64u  djnz    djnzd   djz     djzd    encod   fixinda fixindb
fixindx frac    getacah getacal getacbh getacbl getbyte getcnt
getdivq getdivr getmulh getmull getnib  getnp   getp    getptra
getptrb getptrx getptry getsqrt getword icachen icachep icachex
ijnz    ijnzd   ijz     ijzd    incd    incds   incmod  isob
jmp     jmpd    jmpsw   jmpswd  jnp     jnpd    jnz     jnzd
jp      jpd     jz      jzd     locbase locbyte lockclr locknew
lockret lockset loclong locptra locptrb locword maca    macb
max     maxs    mergew  min     mins    mov     mul     mul32
mul32u  muxc    muxnc   muxnz   muxz    neg     negc    negnc
negnz   negz    not     notb    notp    offp    onecnt  or
pop     popzc   push    pushzc  rcl     rcr     rdaux   rdauxr
rdbyte  rdbytec rdlong  rdlongc rdwide  rdwidec rdword  rdwordc
repd    reps    ret     reta    retad   retb    retbd   retd
retx    retxd   rety    retyd   rev     rol     ror     sar
saracca saraccb saraccs setacca setaccb setb    setbc   setbnc
setbnz  setbyte setbz   setd    seti    setindb setindx setnib
setp    setpc   setpnc  setpnz  setptra setptrb setptrx setptry
setpz   sets    setwide setwidz setword setzc   seussf  seussr
shl     shr     splitw  sqrt32  sqrt64  sub     subabs  subcnt
subptra subptrb subptrx subptry subr    subs    subsx   subx
sumc    sumnc   sumnz   sumz    test    testn   wait    waitcnt
waitpeq waitpne wraux   wrauxr  wrbyte  wrlong  wrwide  wrword
xor     zercnt


Opcodes Not Implemented
-----------------------
addphsa addphsb addphss addpix  bcdbin  binbcd  blnpix  capctra
capctrb capctrs cfgdac0 cfgdac1 cfgdac2 cfgdac3 cfgdacs cfgpins
cmpr    decpat  eswap4  eswap8  getcntx getcosa getcosb getlfsr
getphsa getphsb getphza getphzb getpix  getqx   getqy   getqz
getsina getsinb grybin  incpat  jmplist jmptask locinst mergeb
mixpix  movbyts mulpix  packrgb passcnt polctra polctrb polvid
qartan  qexp    qlog    qrotate qsincos rolbyte rolnib  rolword
scl     serina  serinb  serouta seroutb setbyts setctra setctrb
setctrs setdac0 setdac1 setdac2 setdac3 setdacs setfrqa setfrqb
setfrqs setmap  setmix  setphsa setphsb setphss setpix  setpixa
setpixb setpixg setpixr setpixu setpixv setpixw setpixz setpora
setporb setporc setpord setqi   setqz   setrace setsera setserb
settask setvid  setvidi setvidq setvidy setwava setwavb setwavs
setwrds setx    setxch  setxft  splitb  subphsa subphsb subphss
synctra synctrb taskid  testb   unpkrgb waitpf  waitpr  waitpx
waitvid 


pfth103_p2/changes.txt
                         Changes in pfth 1.03

1. Added umod and u/ words

2. Fixed the #, udmod and ud/ words to handle numbers with the MSB set.
   These changes fix a problem with the u. word when the MSB is set.

3. Fixed the putch PASM routine so that it only writes P30 in OUTA


                         Changes in pfth 1.02

1. Added ospfth.spin, which runs under Spinix

2. Added support for two simultaneous open files

3. Added bye and reboot words

4. Added sub-directory support

5. Added CD word to change the working directory

6. Modified the s" word to work in the interpretation mode

7. Added a kernel word to make constants and values more efficient


                         Changes in pfth 1.00

1. Fixed the FOR and NEXT words

2. Fixed a bug in UNLOOP

3. Added create-file, file-write, file-line, file-flush, file-close and
   delete-file

4. Added linux.fth that implements simple ls, rm, cp and cat commands

5. Added sdutils.fth that contains some file utilities

6. Added TED -- a tiny text editor

7. Changed SAVE to EESAVE in i2c.fth


                         Changes in pfth 0.89

1. Changed the dictionary word structure to use 16-bit pointers instead of 32
   bits.

2. Changed the execution list in compiled words to use 16-bit execution tokens
   instead of 32 bits.

3. Changed the positions of the code pointer and DOES pointer to make the inner
   loop more efficient.  It now runs about 25% faster.

4. Changed SEE back to printing "_lit" for numbers.

5. Implemented a _loop kernel word to make DO-LOOP faster.


                         Changes in pfth 0.83

1. Changed SEE in see.fth so that it doesn't print "_lit" for numbers

2. Added bufser.fth that contains a buffered serial driver

3. Added a BUFFERIT word in bufser.fth that reads the serial input into memory
   until it encounters an "ESC" character.

4. Added an EVALBUF word in bufser.fth that will evaluate multiple lines in a
   memory buffer

5. Updated the readme.txt file with text contributed by Loopy

6. Updated the readme.txt file with a description of the dictionary word format

7. Added a SAVE word to i2c.fth that will write the current hub RAM image to
   EEPROM.  Note: The small Spin program at the end of memory must have been
   preserved for this to boot properly.

8. Changed the CR word so that it emits both a carriage return and a line feed.

9. Added CASE, ENDCASE, OF and ENDOF to init.fth

10. Improved WAITCNT and MS in propwords.fth.  Removed INA!

11. Fixed LIST-FILES in sd.fth to skip over deleted files in the directory

12. Changed the execution token so that it points to the code pointer instead
    of the beginning of the word in the dictionary

13. Made KEY a deferred word so that it can be re-assigned to the serial input
    after reading the FILEs in memory


debug.c
/*******************************************************************************
' Author: Dave Hein
' Version 0.54
' Copyright (c) 2012
' See end of file for terms of use.
'******************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#ifdef LINUX
#include <dirent.h>
#include <sys/stat.h>
#include "conion.h"
#else
#include <conio.h>
#include <direct.h>
#endif
#include "interp.h"
#include "spindebug.h"
#include "eeprom.h"
#include "spinsim.h"

extern int32_t loopcount;
extern int32_t printflag;
extern int32_t baudrate;
extern int32_t eeprom;
extern char *hubram;
extern int32_t printbreak;
extern PasmVarsT PasmVars[8];

void GetDebugString(char *ptr);
int32_t RunProp(int32_t maxloops);

void Help(void)
{
    printf("Debug Commands\n");
    printf("help           - Print command list\n");
    printf("exit           - Exit spinsim\n");
    printf("step           - Run one cycle\n");
    printf("stepx          - Run next executed instruction\n");
    printf("run            - Run continuously\n");
    printf("verbose #      - Set verbosity level\n");
    printf("reboot         - Reboot the Prop\n");
    printf("setbr cog addr - Set breakpoint for cog to addr\n");
    printf("state cog      - Dump cog state\n");
}

char *SkipChar(char *str, int value)
{
    while (*str)
    {
        if (*str != value) break;
        str++;
    }
    return str;
}

void DumpState(PasmVarsT *pasmvars)
{
    printf("cflag = %d, zflag = %d, waitflag = %d\n",
	pasmvars->cflag, pasmvars->zflag, pasmvars->waitflag);
    printf("ptra = %5.5x, ptrb = %5.5x, ptrx = %2.2x, ptry = %2.2x, inda = %3.3x, indb = %3.3x\n",
	pasmvars->ptra, pasmvars->ptra, pasmvars->ptrx,
	pasmvars->ptry, pasmvars->inda, pasmvars->indb);
    printf("pc1 = %8.8x, instruct1 = %8.8x\n", pasmvars->pc1, pasmvars->instruct1);
    printf("pc2 = %8.8x, instruct2 = %8.8x\n", pasmvars->pc2, pasmvars->instruct2);
    printf("pc3 = %8.8x, instruct3 = %8.8x\n", pasmvars->pc3, pasmvars->instruct3);
    printf("pc4 = %8.8x, instruct4 = %8.8x\n", pasmvars->pc4, pasmvars->instruct4);
}

void Debug(void)
{
    int runflag = 1;
    char buffer[200];
    char lastcmd[200];
    int maxloops;
    int stepflag = 0;
    int saveprintflag = 0;

    strcpy(lastcmd, "help");
    while (runflag)
    {
        while (1)
        {
            printf("\nDEBUG> ");
	    fflush(stdout);
            GetDebugString(buffer);
	    if (buffer[0] == 0) strcpy(buffer, lastcmd);
	    strcpy(lastcmd, buffer);
            if (!strcmp(buffer, "exit"))
            {
                runflag = 0;
                break;
            }
            else if (!strcmp(buffer, "step"))
            {
		stepflag = 1;
		saveprintflag = printflag;
		printflag = 0xffffffff;
                LONG(SYS_DEBUG) = printflag;
                maxloops = loopcount + 1;
                break;
            }
            else if (!strcmp(buffer, "stepx"))
            {
		stepflag = 1;
		saveprintflag = printflag;
		printflag = 0x22222222;
                LONG(SYS_DEBUG) = printflag;
		printbreak = 1;
                maxloops = -1;
                break;
            }
            else if (!strncmp(buffer, "verbose", 7))
            {
		char *ptr = SkipChar(buffer+7, ' ');
		if (*ptr == 0)
		    printflag = 0xffffffff;
		else
		{
		    sscanf(ptr, "%x", &printflag);
		    if (ptr[1] == 0)
			printflag *= 0x11111111;
		}
                LONG(SYS_DEBUG) = printflag;
            }
            else if (!strcmp(buffer, "help"))
            {
                Help();
            }
            else if (!strcmp(buffer, "run"))
            {
                maxloops = -1;
                break;
            }
            else if (!strncmp(buffer, "setbr ", 6))
            {
                int cognum, address;
                sscanf(buffer+6, "%x %x", &cognum, &address);
                PasmVars[cognum&7].breakpnt = address;
                LONG(SYS_DEBUG) = printflag;
            }
            else if (!strcmp(buffer, "reboot"))
            {
                RebootProp();
            }
            else if (!strncmp(buffer, "state ", 6))
            {
		int cognum = buffer[6] & 7;
		DumpState(&PasmVars[cognum]);
            }
	    else
		printf("?\n");
        }
        if (runflag) RunProp(maxloops);
	if (stepflag)
	{
	    stepflag = 0;
	    printbreak = 0;
	    printflag = saveprintflag;
	    LONG(SYS_DEBUG) = printflag;
	}
    }
}

void GetDebugString(char *ptr)
{
    int value;

    while (1)
    {
#ifndef STD_CONSOLE_INPUT
        while (!kbhit());
#endif
        value = getch();
        putchx(value);
        if (value == 13 || value == 10)
        {
            *ptr = 0;
            return;
        }
        *ptr++ = value;
    }
}

int32_t RunProp(int32_t maxloops)
{
    int32_t runflag = 1;

    while (runflag && (maxloops < 0 || loopcount < maxloops))
    {
        runflag = step_chip();
        CheckCommand();
        if (baudrate)
        {
            CheckSerialOut();
            if (CheckSerialIn()) return 1;
        }
        if (eeprom)
            CheckEEProm();
    }
    return 0;
}
/*
+------------------------------------------------------------------------------------------------------------------------------+
|                                                   TERMS OF USE: MIT License                                                  |
+------------------------------------------------------------------------------------------------------------------------------+
|Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation    |
|files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy,    |
|modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software|
|is furnished to do so, subject to the following conditions:                                                                   |
|                                                                                                                              |
|The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.|
|                                                                                                                              |
|THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE          |
|WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR         |
|COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,   |
|ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                         |
+------------------------------------------------------------------------------------------------------------------------------+
*/
